app/join_record.ts

```diff
@@ -0,0 +1,127 @@
+"use server";
+// import { access, constants, appendFileSync, readFile } from "node:fs";
+import {
+  S3Client,
+  PutObjectCommand,
+  GetObjectCommand,
+} from "@aws-sdk/client-s3";
+import { JoinFormValues } from "@/components/JoinForm/JoinForm";
+import { Readable } from "stream";
+import { JoinRecord } from "./types";
+
+// const JOIN_FORM_LOG = process.env.JOIN_FORM_LOG as string;
+
+class JoinRecordS3 {
+  private s3Client: S3Client;
+
+  private S3_REGION: string;
+  private S3_ENDPOINT: string;
+  private S3_BUCKET_NAME: string;
+  private S3_BASE_NAME: string;
+  private S3_ACCESS_KEY: string;
+  private S3_SECRET_KEY: string;
+
+  constructor() {
+    this.S3_REGION = process.env.S3_REGION as string;
+    this.S3_ENDPOINT = process.env.S3_ENDPOINT as string;
+    this.S3_BUCKET_NAME = process.env.S3_BUCKET_NAME as string;
+    this.S3_BASE_NAME = process.env.S3_BASE_NAME as string;
+    this.S3_ACCESS_KEY = process.env.S3_ACCESS_KEY as string;
+    this.S3_SECRET_KEY = process.env.S3_SECRET_KEY as string;
+
+    // Setup the S3 client
+    this.s3Client = new S3Client({
+      region: this.S3_REGION != undefined ? this.S3_REGION : "us-east-1",
+      endpoint:
+        this.S3_ENDPOINT != undefined
+          ? this.S3_ENDPOINT
+          : "https://s3.us-east-1.amazonaws.com",
+      credentials: {
+        accessKeyId: this.S3_ACCESS_KEY,
+        secretAccessKey: this.S3_SECRET_KEY,
+      },
+    });
+  }
+
+  // Records the submission we just got as a JSON object in an S3 bucket.
+  // submission: A Join Form Submission. We append a few things to this.
+  // key: The S3 path we store the submission at
+  // responseCode: If we have a response code for this submission, add it here.
+  async save(joinRecord: JoinRecord, key: string = "") {
+    // Bail if there's no S3 key
+    if (this.S3_ACCESS_KEY === undefined || this.S3_SECRET_KEY === undefined) {
+      console.error(
+        "S3 credentials not configured. I WILL NOT SAVE THIS SUBMISSION.",
+      );
+      return;
+    }
+
+    // Get the date to store this submission under (this is part of the path)
+    const submissionKey = joinRecord.submission_time
+      .replace(/[-:T]/g, "/")
+      .slice(0, 19);
+
+    // Create the path, or use the one provided.
+    key = key != "" ? key : `${this.S3_BASE_NAME}/${submissionKey}.json`;
+
+    let body = JSON.stringify(joinRecord);
+
+    const command = new PutObjectCommand({
+      Bucket: this.S3_BUCKET_NAME,
+      Key: key,
+      Body: body,
+    });
+
+    try {
+      const response = await this.s3Client.send(command);
+      console.log(response);
+    } catch (err) {
+      // Oof, guess we'll drop this on the floor.
+      console.error(err);
```

Comment:
Can we set up some kind of datadog alert for this at least?

Reply:
We can definitely emit a metric. TBH I was trying to avoid setting up metrics for as long as possible 😂